rust/src/records.rs

```diff
@@ -0,0 +1,267 @@
+//! Struct definitions for all [MCAP Records](https://mcap.dev/specification/index.html#records).
+use lifetime::IntoStatic;
+use serde::{Deserialize, Serialize};
+use std::borrow::Cow;
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
+
+pub type Timestamp = u64;
+
+#[derive(Debug, Copy, PartialEq, Eq, Serialize, Deserialize, Clone)]
+pub enum OpCode {
+    Header,
+    Footer,
+    Schema,
+    Channel,
+    Message,
+    Chunk,
+    MessageIndex,
+    ChunkIndex,
+    Attachment,
+    AttachmentIndex,
+    Statistics,
+    Metadata,
+    MetadataIndex,
+    SummaryOffset,
+    DataEnd,
+    UserOpcode(u8),
+}
+
+pub struct InvalidOpcode(pub u8);
+
+impl TryFrom<u8> for OpCode {
+    type Error = InvalidOpcode;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0x00 => Err(InvalidOpcode(value)),
+            0x01 => Ok(OpCode::Header),
+            0x02 => Ok(OpCode::Footer),
+            0x03 => Ok(OpCode::Schema),
+            0x04 => Ok(OpCode::Channel),
+            0x05 => Ok(OpCode::Message),
+            0x06 => Ok(OpCode::Chunk),
+            0x07 => Ok(OpCode::MessageIndex),
+            0x08 => Ok(OpCode::ChunkIndex),
+            0x09 => Ok(OpCode::Attachment),
+            0x0A => Ok(OpCode::AttachmentIndex),
+            0x0B => Ok(OpCode::Statistics),
+            0x0C => Ok(OpCode::Metadata),
+            0x0D => Ok(OpCode::MetadataIndex),
+            0x0E => Ok(OpCode::SummaryOffset),
+            0x0F => Ok(OpCode::DataEnd),
+            x if x < 0x80 => Err(InvalidOpcode(x)),
```

Comment:
good idea, we should probably add this check and the UserOpcode equivalent in other libraries. I don't think we have that today